### ISSUE TITLE
Scheduler "deadlocks" itself when max_active_runs_per_dag is reached by up_for_retry tasks

### ISSUE THREAD

**Issue by antontimenko** [apache/airflow#14205]
**Apache Airflow version**: 2.0.1

**What happened**:

Let's say we have `max_active_runs_per_dag = 2` in config. Now we manually trigger, for example, 10 DAG runs for some specific DAG. In the DAG there are some tasks, that should be retried on fail with some interval.

The issue is when at least 2 DAG runs have tasks inside that are failed, moved to up_for_retry state, and waiting to be rescheduled again, the scheduler will not reschedule them at all. In stdout it keeps saying that `DAG <dag_name> already has 2 active DAG runs, not queuing any tasks for run <execution_date>`. Even DAG runs inside other DAGs stop to run

Executor: CeleryExecutor

**What you expected to happen**:

I expected that up_for_retry tasks would be rescheduled when they reached their retry interval

**How to reproduce it**:

Just follow the instructions above. Set `max_active_runs_per_dag = 2`, create a DAG with PythonOperator with the function inside that fails, set retry_delay to something like 1 minute, trigger manually 2 DAG runs, and verify that task wouldn't be rescheduled on delay

**Comment by yergi** [comment]
I think the issue I'm experiencing is related to this. 

Apache Airflow version: 2.0.1
Executor: LocalExecutor

What happened: I have `max_active` set to 4, and when running a backfill for this dag, if 4 sensor tasks get set for `up_for_reschedule` at the same time, the backfill exits telling me that all the tasks downstream for these sensors are deadlocked.

**Comment by zachliu** [comment]
if https://github.com/apache/airflow/issues/9975 is fixed, does this one fixes itself too? because you'll never have more runs than specified by the `max_active_runs`

**Comment by antontimenko** [comment]
@zachliu I think that this problem wouldn't be fixed because this occurs even when number of active runs == `max_active_runs `. So if `max_active_runs` == 2, and there is 2 `up_for_retry` tasks, it will deadlock itself

**Comment by zachliu** [comment]
@antontimenko ah i see, i have `max_active_runs_per_dag = 16`. and i rarely reach `16` task runs per dag. but when https://github.com/apache/airflow/issues/9975 occurs, the `up_for_retry` tasks can easily go over `16` and then my scheduler got `deadlocked` :joy_cat:

**Comment by zachliu** [comment]
@antontimenko i think i kinda understand your situation :thinking: 

so you're using a large number of DAG runs to communicate with some unreliable external APIs
you spin up 7000 dag runs in order to query the said APIs with different parameters

i have a similar situation. but instead of thousands of dag runs, i have thousands of processes (python multiprocessing)

one thing i don't quite understand is that if `these DAG runs are created using some kind of code that makes requests to Airflow HTTP API and triggers these runs`, can't you control it somehow? e.g. pacing the code so that it doesn't create more than 16 running tasks

**Comment by antontimenko** [comment]
Of course, I can spend some time creating some kind of logic, that constantly makes requests to Airflow API to get the number of active DAG runs and trigger new ones only when there are less than 16 of them. But this would look like I'm trying to create some kind of my own scheduler on top of Airflow scheduler, just to get around this bug :)

I think that scheduler should process any kind of DAG runs properly, even if there are more of them than `max_active_runs`. And as I said above, it worked properly in earlier versions

**Comment by zachliu** [comment]
fair enough, i can see this is almost like a deal breaker for you :crying_cat_face:

**Comment by zachliu** [comment]
@antontimenko we've been talking here without touching the most important point :joy_cat:  is there a quick fix when this issue occurs?
* restarting the scheduler probably won't do anything right?
* clearing the `up_for_retry` tasks and wait for scheduler to deadlock itself again (and keep doing this until the # of `up_for_retry` tasks is less than `max_active_runs_per_dag`)

**Comment by antontimenko** [comment]
@zachliu Currently the fix is to clear `up_for_retry` tasks or run them manually from UI when this occurs. Restarting scheduler doesn't help, yes

**Comment by sosso** [comment]
+1 for us on this issue as well, I think?  Very strangely, we see the most recent run for a DAG have its *run* be set to 'running', but the only task in the DAG be a clear success:

![image](https://user-images.githubusercontent.com/619968/115061689-57345180-9e9e-11eb-88a4-41de2abf94d2.png)

This is a catchup=False DAG, where the only task runs in a pool, and there is *nothing* in the Scheduler for this DAG for two hours (the DAG runs is supposed to run every 5 minutes) about why it can't schedule this DAG.  No "max active runs reached", no "no slots available in pool", nothing.  It's like the scheduler forgot this DAG existed until we rebooted it.

**Comment by xiskoct** [comment]
the same thing happens to me, huge problem.

**Comment by sosso** [comment]
After a few hours spelunking through the codebase, we came across `max_dagruns_per_loop_to_schedule` in `scheduler_job.py`.  We have several thousand DAGs, all between 1-3 tasks (80% are probably 1 task) and this was a massive improvement for us in setting it to 200 (10x the default value of 20).

**Comment by zachliu** [comment]
I forgot to ask this simple question:
Since the scheduler is now HA compatible in 2.x, does it help if we spin up multiple schedulers?
I'm hoping if one got deadlocked, the other one is still alive :grin:

**Comment by yuqian90** [comment]
The symptom I'm seeing is that whenever any single DAG hits the `max_active_runs_per_dag` limit, the scheduler log will be full of lines like this. At this point, the scheduler basically stops scheduling any task for any dags. The fact that one dag hitting this limit causes all other dags to be halted is pretty bad. Our workaround at this moment is to make `max_active_runs_per_dag` a big number that is unlikely going to be hit.

```
DAG ... already has .. active runs, not queuing any tasks
```

**Comment by richddr** [comment]
Also seeing this issue, @antontimenko thanks for surfacing!

**Comment by richddr** [comment]
@zachliu +1 to what @yuqian90 said. On our configuration we're running multiple schedulers and still see the same issue.

**Comment by zachliu** [comment]
> @zachliu +1 to what @yuqian90 said. On our configuration we're running multiple schedulers and still see the same issue.

oh no, all schedulers are `deadlocked` at the same time!? 😓

**Comment by richddr** [comment]
@zachliu unfortunately yes 😭

**Comment by ashb** [comment]
I wonder if this will already be fixed by Ephraim's change where DagRuns are created in queued state - #16854

**Comment by tzanko-matev** [comment]
Here is what I believe is causing the bug: 
https://github.com/apache/airflow/blob/88199eefccb4c805f8d6527bab5bf600b397c35e/airflow/jobs/scheduler_job.py#L1765

```py
        if dag.max_active_runs:
            if (
                len(currently_active_runs) >= dag.max_active_runs
                and dag_run.execution_date not in currently_active_runs
            ):
                self.log.info(
                    "DAG %s already has %d active runs, not queuing any tasks for run %s",
                    dag.dag_id,
                    len(currently_active_runs),
                    dag_run.execution_date,
                )
                return 0
```
Notice that the dagrun state is not updated before `return 0`. This means that on the next iteration of the scheduler the same dagrun will be tried for processing. If there are enough dagruns which satisfy this condition then the scheduler will never try to process other dagruns and will deadlock.

The code which makes the scheduler go forward is found in `dag_run.update_state`:
https://github.com/apache/airflow/blob/88199eefccb4c805f8d6527bab5bf600b397c35e/airflow/models/dagrun.py#L384

```py
    @provide_session
    def update_state(
        self, session: Session = None, execute_callbacks: bool = True
    ) -> Tuple[List[TI], Optional[callback_requests.DagCallbackRequest]]:
        
        #..........
        start_dttm = timezone.utcnow()
        self.last_scheduling_decision = start_dttm
        # ........
        session.merge(self)

        return schedulable_tis, callback
```
Updating `last_scheduling_decision` is what makes the scheduler chose a different set of dagruns on the next iteration. To fix the bug this update needs to be made before the `return 0` line in the first code snippet.

In the current main branch the scheduling logic is different, so the bug might have disappeared. The bug exists in all versions up to 2.1.2.

**Comment by tzanko-matev** [comment]
Here is the same logic in the `main` branch: 
https://github.com/apache/airflow/blob/721d4e7c60cbccfd064572f16c3941f41ff8ab3a/airflow/jobs/scheduler_job.py#L1004
```py
            if dag.max_active_runs and active_runs >= dag.max_active_runs:
                self.log.debug(
                    "DAG %s already has %d active runs, not moving any more runs to RUNNING state %s",
                    dag.dag_id,
                    active_runs,
                    dag_run.execution_date,
                )
```
After a brief look at the code (without testing) it seems to me that the same problem will be happening. Since `last_scheduling_decision` is not updated you might end up with the same type of deadlock because you will be examining the same batch of dagruns in state `QUEUED` again and again and you won't be able to move any of them to state `RUNNING`.

**Comment by ephraimbuddy** [comment]
I have made a PR related to this issue, see https://github.com/apache/airflow/pull/17945

What happens is that the method `DagRun.next_dagruns_to_examine` gets the earliest dagruns without considering the dag that has the dagrun.
For example:
If you have a dag with execution_date 2020,1,1 and set catchup=True, max_active_runs=1, schedule_interval='@daily' and another dag with execution_date 2021,1,1 and also set catchup=True,  schedule_interval='@daily'.
When you unpause the two dags, the dagruns would be created but only one dagrun would be active because of how  `DagRun.next_dagruns_to_examine` works.
I'm hopeful my PR would resolve this issue but I'm worried about performance. Please take a look: https://github.com/apache/airflow/pull/17945 @uranusjr @kaxil @ash

**Comment by kaxil** [comment]
Can this be closed @ephraimbuddy ???

**Comment by kaxil** [comment]
I am hoping this is solved by https://github.com/apache/airflow/pull/17945 which will be released in 2.1.4 (should be released tomorrow). We can reopen if this is still an issue

**Comment by potiuk** [comment]
🤞
